src/server/endpoints/covidcast_utils/trend.py

```diff
@@ -62,6 +65,54 @@ def compute_trend(geo_type: str, geo_value: str, signal_source: str, signal_sign
     return t
 
 
+def compute_trends(geo_type: str, geo_value: str, signal_source: str, signal_signal: str, shifter: Callable[[int], int], rows: Iterable[Tuple[int, float]]) -> List[Trend]:
+    min_value: Optional[float] = None
+    min_date: Optional[int] = None
+    max_value: Optional[float] = None
+    max_date: Optional[int] = None
+
+    lookup: Dict[int, float] = OrderedDict()
+    # find all needed rows
+    for time, value in rows:
+        lookup[time] = value
+        if min_value is None or min_value > value:
+            min_date = time
+            min_value = value
+        if max_value is None or max_value < value:
+            max_date = time
+            max_value = value
+
+    trends: List[Trend] = []
+    for current_time, value in lookup.items():
+        basis_time = shifter(current_time)
+        basis_value = lookup.get(basis_time, None)
+        t = Trend(
+            geo_type,
+            geo_value,
+            signal_source,
+            signal_signal,
+            date=current_time,
+            value=value,
+            basis_date=basis_time if basis_value is not None else None,
+            basis_value=basis_value,
+            min_date=min_date,
+            min_value=min_value,
+            max_date=max_date,
+            max_value=max_value,
+        )
+
+        trends.append(t)
+
+        if t.value is None or t.min_value is None:
+            continue
+
+        t.basis_trend = compute_trend_class(compute_trend_value(t.value, t.basis_value, t.min_value)) if t.basis_value else TrendEnum.unknown
+        t.min_trend = compute_trend_class(compute_trend_value(t.value, t.min_value, t.min_value))
+        t.max_trend = compute_trend_class(compute_trend_value(t.value, t.max_value, t.min_value)) if t.max_value else TrendEnum.unknown
```

Comment:
is the downstream code that makes use of these attributes aware that they are not always set?

Reply:
yes, see also interface on the www-covidcast side: https://github.com/cmu-delphi/www-covidcast/blob/9ba9f934aa37c24c936cb06374f203f061cebd1e/src/data/api.ts#L115-L136